sklearn/preprocessing/imputation.py

```diff
@@ -374,3 +374,211 @@ def transform(self, X):
+        The features with missing values.
+
+    n_features_ : int
+        The number of features in the input.
+    """
+
+    def __init__(self, missing_values=np.NaN, features="train", sparse="auto", strict=True):
+        self.missing_values = missing_values
+        self.features = features
+        self.sparse = sparse
+        self.strict = strict
+
+    def fit(self, X, y=None):
+        """Fit the transformer on X.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape (n_samples, n_features)
+            Input data, where ``n_samples`` is the number of samples and
+            ``n_features`` is the number of features.
+
+        Returns
+        -------
+        self : object
+            Returns self.
+        """
+        self._validate_params()
+
+        X = check_array(X, accept_sparse=('csc', 'csr'), dtype=np.float64, force_all_finite=False)
+        self.n_features_ = X.shape[1]
+
+        if isinstance(self.features, six.string_types):
+            if self.features == "train":
+                _, self.feat_with_missing_ = self._get_missing_features_info(X)
+            else : # self.features == "all"
+                self.feat_with_missing_ = np.arange(self.n_features_)
+        else:
+            self.feat_with_missing_ = self.features
+
+        return self
+
+    def transform(self, X):
+        """Generate missing values indicator for X.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape = [n_samples, n_features]
+            The input data to complete.
+
+        Returns
+        -------
+        Xt : array or sparse matrix, shape = [n_samples, n_features]
+             The missing indicator for input data
+
+        """
+        check_is_fitted(self, "feat_with_missing_", "n_features_")
+
+        X = check_array(X, accept_sparse=('csc', 'csr'), dtype=np.float64, force_all_finite=False)
+        if X.shape[1] != self.n_features_:
+            raise ValueError("X has a different number of features "
+                             "than during fitting.")
+
+        imputer_mask, feat_with_missing = self._get_missing_features_info(X)
+
+        if isinstance(self.features, six.string_types):
+            if self.features == "train":
+                features = np.setdiff1d(feat_with_missing,
+                                        self.feat_with_missing_)
+                if self.strict and features.size > 0:
+                    raise Exception("The features %s have missing values "
+                                    "in transform but have no missing values "
+                                    "in fit" % features)
+
+        if not (isinstance(self.features, six.string_types) and self.features == "all"):
+            # no need to slice when all features have missing values
+            imputer_mask = imputer_mask[:, self.feat_with_missing_]
+
+        return imputer_mask
+
+    def fit_transform(self, X, y=None):
```

Comment:
are we gaining anything from this apart from avoiding doing ``check_array`` twice? We have ``force_all_finite=False`` so it shouldn't be that slow, right? This looks like a lot of duplicate code.